demos/signature/EVP_DSA_Signature_demo.c

```diff
@@ -0,0 +1,315 @@
+    EVP_MD_CTX *ctx = NULL;
+    EVP_PKEY_CTX *pkey_ctx = NULL;
+    EVP_PKEY *pkey = NULL;
+
+    pkey_ctx = EVP_PKEY_CTX_new_from_name(libctx, ALG, PROPQUERY);
+    if (pkey_ctx == NULL)
+        goto end;
+    if (EVP_PKEY_fromdata_init(pkey_ctx) != 1)
+        goto end;
+    if (EVP_PKEY_fromdata(pkey_ctx, &pkey, EVP_PKEY_KEYPAIR, keypair) != 1)
+        goto end;
+
+    ctx = EVP_MD_CTX_create();
+    if (ctx == NULL)
+        goto end;
+
+    if (EVP_DigestSignInit_ex(ctx, NULL, DIGEST, libctx, NULL, pkey, NULL) != 1)
+        goto end;
+
+    if (EVP_DigestSignUpdate(ctx, hamlet_1, sizeof(hamlet_1)) != 1)
+        goto end;
+
+    if (EVP_DigestSignUpdate(ctx, hamlet_2, sizeof(hamlet_2)) != 1)
+        goto end;
+
+    /* Calculate the signature size */
+    if (EVP_DigestSignFinal(ctx, NULL, &sig_len) != 1)
+        goto end;
+    if (sig_len == 0)
+        goto end;
+
+    sig_value = OPENSSL_malloc(sig_len);
+    if (sig_value == NULL)
+        goto end;
+
+    /* Calculate the signature */
+    if (EVP_DigestSignFinal(ctx, sig_value, &sig_len) != 1)
+        goto end;
+
+    result = 1;
+end:
+    EVP_MD_CTX_free(ctx);
+    if (result != 1) {
+        OPENSSL_free(sig_value);
+        sig_len = 0;
+        sig_value = NULL;
+    }
+    *p_sig_len = sig_len;
+    *p_sig_value = sig_value;
+    EVP_PKEY_free(pkey);
+    EVP_PKEY_CTX_free(pkey_ctx);
+
+    fprintf(stdout, "Generating signature:\n");
+    BIO_dump_indent_fp(stdout, sig_value, sig_len, 2);
+    fprintf(stdout, "\n");
+    return result;
+}
+
+static int demo_verify(OSSL_LIB_CTX *libctx,
+                       size_t sig_len, unsigned char *sig_value,
+                       OSSL_PARAM public_key[])
+{
+    int result = 0;
+    EVP_MD_CTX *ctx = NULL;
+    EVP_PKEY_CTX *pkey_ctx = NULL;
+    EVP_PKEY *pkey = NULL;
+
+    pkey_ctx = EVP_PKEY_CTX_new_from_name(libctx, ALG, PROPQUERY);
+    if (pkey_ctx == NULL)
+        goto end;
+    if (EVP_PKEY_fromdata_init(pkey_ctx) != 1)
+        goto end;
+    if (EVP_PKEY_fromdata(pkey_ctx, &pkey, EVP_PKEY_PUBLIC_KEY, public_key) != 1)
+        goto end;
+
+    ctx = EVP_MD_CTX_create();
+    if(ctx == NULL)
+        goto end;
+
+    if (EVP_DigestVerifyInit_ex(ctx, NULL, DIGEST, libctx, NULL, pkey, NULL) != 1)
+        goto end;
+
+    if (EVP_DigestVerifyUpdate(ctx, hamlet_1, sizeof(hamlet_1)) != 1)
+        goto end;
+
+    if (EVP_DigestVerifyUpdate(ctx, hamlet_2, sizeof(hamlet_2)) != 1)
+        goto end;
+
+    if (EVP_DigestVerifyFinal(ctx, sig_value, sig_len) != 1)
+        goto end;
+
+    result = 1;
+end:
+    EVP_PKEY_free(pkey);
+    EVP_PKEY_CTX_free(pkey_ctx);
+    EVP_MD_CTX_free(ctx);
+    return result;
+}
+
+int main(void)
+{
+    int result = 0;
+    OSSL_LIB_CTX *libctx = NULL;
+    EVP_PKEY *params = NULL;
+    EVP_PKEY *pkey = NULL;
+    OSSL_PARAM *public_key = NULL;
+    OSSL_PARAM *keypair = NULL;
+    size_t sig_len = 0;
+    unsigned char *sig_value = NULL;
+
+    libctx = OSSL_LIB_CTX_new();
```

Comment:
check for null here also. If this failed it would fallback to using the default libctx